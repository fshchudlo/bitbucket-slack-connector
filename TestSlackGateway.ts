import * as slack from "@slack/web-api";
import { SlackGateway } from "./contracts";

export class TestSlackGateway implements SlackGateway {
    snapshot: {
        createdChannels: slack.ConversationsCreateArguments[];
        setChannelTopics: slack.ConversationsSetTopicArguments[];
        invitesToChannels: slack.ConversationsInviteArguments[];
        closedChannels: slack.ConversationsCloseArguments[];
        sentMessages: slack.ChatPostMessageArguments[];
        lookedUpUsers: slack.UsersLookupByEmailArguments[];
    };
    constructor() {
        this.snapshot = {
            createdChannels: new Array<slack.ConversationsCreateArguments>(),
            setChannelTopics: new Array<slack.ConversationsSetTopicArguments>(),
            invitesToChannels: new Array<slack.ConversationsInviteArguments>(),
            closedChannels: new Array<slack.ConversationsCloseArguments>(),
            sentMessages: new Array<slack.ChatPostMessageArguments>(),
            lookedUpUsers: new Array<slack.UsersLookupByEmailArguments>(),
        };
    }
    lookupUserByEmail(options: slack.UsersLookupByEmailArguments): Promise<slack.UsersLookupByEmailResponse> {
        this.snapshot.lookedUpUsers.push(options);
        return Promise.resolve({ ok: true, user: { id: options.email } });
    }
    createChannel(options: slack.ConversationsCreateArguments): Promise<slack.ConversationsCreateResponse> {
        this.snapshot.createdChannels.push(options);
        return Promise.resolve({ ok: true, channel: { id: "12345", name: options.name } });
    }
    setChannelTopic(options: slack.ConversationsSetTopicArguments): Promise<slack.ConversationsSetTopicResponse> {
        this.snapshot.setChannelTopics.push(options);
        return Promise.resolve({ ok: true });
    }
    inviteToChannel(options: slack.ConversationsInviteArguments): Promise<slack.ConversationsInviteResponse> {
        this.snapshot.invitesToChannels.push(options);
        return Promise.resolve({ ok: true });
    }
    closeChannel(options: slack.ConversationsCloseArguments): Promise<slack.ConversationsCloseResponse> {
        this.snapshot.closedChannels.push(options);
        return Promise.resolve({ ok: true });
    }
    sendMessage(options: slack.ChatPostMessageArguments): Promise<slack.ChatPostMessageResponse> {
        this.snapshot.sentMessages.push(options);
        return Promise.resolve({ ok: true });
    }
}
