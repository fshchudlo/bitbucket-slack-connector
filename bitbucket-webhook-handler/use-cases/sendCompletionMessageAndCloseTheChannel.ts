import { iconEmoji } from "./slack-building-blocks";
import { SendMessageArguments, SlackNotificationChannel } from "../SlackNotificationChannel";
import { PullRequestBasicNotification } from "../../bitbucket-payload-types";
import { getPullRequestCompletionAction } from "./helpers/getPullRequestCompletionAction";

export async function sendCompletionMessageAndCloseTheChannel(payload: PullRequestBasicNotification, slackAPI: SlackNotificationChannel, slackChannelId: string) {
    await slackAPI.sendMessage(buildMessage(payload, slackChannelId));
    await slackAPI.closeChannel(slackChannelId);
}


function buildMessage(payload: PullRequestBasicNotification, channelId: string): SendMessageArguments {
    const completionAction = getPullRequestCompletionAction(payload);
    return {
        channelId: channelId,
        iconEmoji: iconEmoji,
        text: `${completionAction.emoji} ${completionAction.text}`
    };
}