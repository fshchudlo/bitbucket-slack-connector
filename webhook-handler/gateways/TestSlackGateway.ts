import * as slack from "@slack/web-api";

import { SlackGateway } from "./SlackGateway";

const channelId = "12345";
export default class TestSlackGateway implements SlackGateway {
    snapshot: {
        createdChannels: slack.ConversationsCreateArguments[];
        setChannelTopics: slack.ConversationsSetTopicArguments[];
        invitesToChannels: slack.ConversationsInviteArguments[];
        archivedChannels: slack.ConversationsCloseArguments[];
        sentMessages: slack.ChatPostMessageArguments[];
        lookedUpUsers: slack.UsersLookupByEmailArguments[];
    };

    constructor() {
        this.snapshot = {
            createdChannels: new Array<slack.ConversationsCreateArguments>(),
            setChannelTopics: new Array<slack.ConversationsSetTopicArguments>(),
            invitesToChannels: new Array<slack.ConversationsInviteArguments>(),
            archivedChannels: new Array<slack.ConversationsCloseArguments>(),
            sentMessages: new Array<slack.ChatPostMessageArguments>(),
            lookedUpUsers: new Array<slack.UsersLookupByEmailArguments>()
        };
    }

    lookupUserByEmail(options: slack.UsersLookupByEmailArguments): Promise<slack.UsersLookupByEmailResponse> {
        this.snapshot.lookedUpUsers.push(options);
        return Promise.resolve({ ok: true, user: { id: options.email } });
    }

    createChannel(options: slack.ConversationsCreateArguments): Promise<slack.ConversationsCreateResponse> {
        this.snapshot.createdChannels.push(options);
        return Promise.resolve({ ok: true, channel: { id: channelId, name: options.name } });
    }

    setChannelTopic(options: slack.ConversationsSetTopicArguments): Promise<slack.ConversationsSetTopicResponse> {
        this.snapshot.setChannelTopics.push(options);
        return Promise.resolve({ ok: true });
    }

    inviteToChannel(options: slack.ConversationsInviteArguments): Promise<slack.ConversationsInviteResponse> {
        this.snapshot.invitesToChannels.push(options);
        return Promise.resolve({ ok: true });
    }

    archiveChannel(options: slack.ConversationsArchiveArguments): Promise<slack.ConversationsArchiveResponse> {
        this.snapshot.archivedChannels.push(options);
        return Promise.resolve({ ok: true });
    }

    sendMessage(options: slack.ChatPostMessageArguments): Promise<slack.ChatPostMessageResponse> {
        this.snapshot.sentMessages.push(options);
        return Promise.resolve({ ok: true, channel: channelId });
    }
}
