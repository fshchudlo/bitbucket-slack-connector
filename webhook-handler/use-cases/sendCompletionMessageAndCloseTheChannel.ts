import { buildChannelName, formatUserName, iconEmoji } from "../slack-building-blocks";
import { SlackAPIAdapter } from "../SlackAPIAdapter";
import { PullRequestBasicNotification } from "../../typings";

export async function sendCompletionMessageAndCloseTheChannel(payload: PullRequestBasicNotification, slackGateway: SlackAPIAdapter) {
    const channelId = (await slackGateway.sendMessage(buildMessage(payload))).channel;
    await slackGateway.archiveChannel(channelId);
}

function buildMessage(payload: PullRequestBasicNotification) {
    const channelName = buildChannelName(payload.pullRequest);
    let messageText = null;
    switch (payload.eventKey) {
        case "pr:deleted":
            messageText = `:no_entry_sign: The pull request was deleted by ${formatUserName(payload.actor)}.`;
            break;
        case "pr:merged":
            messageText = `:white_check_mark: The pull request was merged by ${formatUserName(payload.actor)}. Well done, thank you all.`;
            break;
        case "pr:declined":
            messageText = `:no_entry_sign: The pull request was declined by ${formatUserName(payload.actor)}.`;
            break;
    }
    return {
        channel: channelName,
        icon_emoji: iconEmoji,
        text: messageText
    };
}
