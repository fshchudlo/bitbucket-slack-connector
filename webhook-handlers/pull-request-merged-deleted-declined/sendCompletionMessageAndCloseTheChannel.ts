import buildChannelName from "../buildChannelName";

export async function sendCompletionMessageAndCloseTheChannel(
    payload: PullRequestMergedDeclinedDeletedPayload,
    slackGateway: SlackGateway
) {
    const pullRequest = payload.pullRequest;
    const channelName = buildChannelName(pullRequest.toRef.repository.project.key, pullRequest.toRef.repository.slug, pullRequest.id);
    let message = null;
    switch (payload.eventKey) {
        case "pr:deleted":
            message = `The pull request was deleted by ${payload.actor.name}.`;
            break;
        case "pr:merged":
            message = `The pull request was merged by ${payload.actor.name}. Well done, thank you all.`;
            break;
        case "pr:declined":
            message = `The pull request was declined by ${payload.actor.name}.`;
            break;
        default:
            throw new Error("Unknown payload type.");
    }

    await slackGateway.sendMessage({
        channel: channelName,
        text: message,
    });

    await slackGateway.closeChannel({
        channel: channelName,
    });
}
