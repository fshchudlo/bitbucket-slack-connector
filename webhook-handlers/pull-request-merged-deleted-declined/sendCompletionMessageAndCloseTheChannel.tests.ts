// mainApp.test.ts
import { TestSlackGateway } from "../../TestSlackGateway";
import { sendCompletionMessageAndCloseTheChannel } from "./sendCompletionMessageAndCloseTheChannel";

let payload: PullRequestMergedDeclinedDeletedPayload = {
    eventKey: "pr:merged",
    date: "2017-09-19T09:58:11+1000",
    actor: {
        name: "test.author",
        emailAddress: "test.author@test.com",
    },
    pullRequest: {
        id: 1,
        title: "Quite a long and comprehensive description to get channel title longer than 250 symbols which is a maximum length for the Slack",
        toRef: {
            displayId: "master",
            repository: {
                slug: "test-repository",
                project: {
                    key: "TEST-PROJ",
                    name: "Test project",
                },
            },
        },
        author: {
            user: {
                name: "test.author",
                emailAddress: "test.author@test.com",
            },
        },
        reviewers: [
            {
                user: {
                    name: "test.reviewer1",
                    emailAddress: "test.reviewer1@test.com",
                },
            },
        ],
        links: {
            self: ["https://git.test.com/projects/TEST-PROJ/repos/test-repository/pull-requests/1/overview"],
        },
    },
};

describe("sendCompletionMessageAndCloseTheChannel", () => {
    it("Should send completion message and close the channel on PR merge", async () => {
        const testee = new TestSlackGateway();
        payload.eventKey = "pr:merged";
        await sendCompletionMessageAndCloseTheChannel(payload, testee);

        expect(testee.snapshot).toMatchSnapshot();
    });

    it("Should send completion message and close the channel on PR declining", async () => {
        const testee = new TestSlackGateway();
        payload.eventKey = "pr:declined";

        await sendCompletionMessageAndCloseTheChannel(payload, testee);

        expect(testee.snapshot).toMatchSnapshot();
    });

    it("Should send completion message and close the channel on PR deletion", async () => {
        const testee = new TestSlackGateway();
        payload.eventKey = "pr:deleted";

        await sendCompletionMessageAndCloseTheChannel(payload, testee);

        expect(testee.snapshot).toMatchSnapshot();
    });

    it("Should throw Error on unkonwn action type", async () => {
        const testee = new TestSlackGateway();
        payload.eventKey = "unknown action";

        expect.assertions(1);
        try {
            await sendCompletionMessageAndCloseTheChannel(payload, testee);
        } catch (error) {
            expect((error as Error).message).toBe("Unknown payload type.");
        }
    });
});
