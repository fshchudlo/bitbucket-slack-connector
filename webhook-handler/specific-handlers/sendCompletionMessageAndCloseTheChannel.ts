import buildChannelName from "../helper-functions/buildChannelName";
import { SlackGateway } from "../gateways/SlackGateway";
import { formatUserName } from "../slack-building-blocks/formatUserName";
import { PullRequestBasicNotification } from "../../typings";
import { getMessageColor } from "../slack-building-blocks/getMessageColor";

export async function sendCompletionMessageAndCloseTheChannel(
    payload: PullRequestBasicNotification,
    slackGateway: SlackGateway,
    iconEmoji: string
) {
    const pullRequest = payload.pullRequest;
    let message = null;
    switch (payload.eventKey) {
        case "pr:deleted":
            message = `The pull request was deleted by ${formatUserName(payload.actor)}.`;
            break;
        case "pr:merged":
            message = `The pull request was merged by ${formatUserName(payload.actor)}. Well done, thank you all.`;
            break;
        case "pr:declined":
            message = `The pull request was declined by ${formatUserName(payload.actor)}.`;
            break;
    }

    const messageResponse = await slackGateway.sendMessage({
        channel: buildChannelName(pullRequest),
        icon_emoji: iconEmoji,
        attachments: [
            {
                text: message,
                color: getMessageColor(payload)
            }]
    });

    await slackGateway.archiveChannel({
        //Close channel is quite unique and requires channel id, not the name. Take that id from sendMessage response to avoid extra triggers of Slack API
        channel: messageResponse.channel
    });
}
