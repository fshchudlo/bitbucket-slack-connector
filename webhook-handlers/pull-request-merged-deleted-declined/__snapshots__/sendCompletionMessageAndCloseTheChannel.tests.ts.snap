// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`sendCompletionMessageAndCloseTheChannel Should send completion message and close the channel on PR declining 1`] = `
{
  "closedChannels": [
    {
      "channel": "pr-test-proj-test-repository-1",
    },
  ],
  "createdChannels": [],
  "invitesToChannels": [],
  "lookedUpUsers": [],
  "sentMessages": [
    {
      "channel": "pr-test-proj-test-repository-1",
      "text": "The pull request was declined by test.author.",
    },
  ],
  "setChannelTopics": [],
}
`;

exports[`sendCompletionMessageAndCloseTheChannel Should send completion message and close the channel on PR deletion 1`] = `
{
  "closedChannels": [
    {
      "channel": "pr-test-proj-test-repository-1",
    },
  ],
  "createdChannels": [],
  "invitesToChannels": [],
  "lookedUpUsers": [],
  "sentMessages": [
    {
      "channel": "pr-test-proj-test-repository-1",
      "text": "The pull request was deleted by test.author.",
    },
  ],
  "setChannelTopics": [],
}
`;

exports[`sendCompletionMessageAndCloseTheChannel Should send completion message and close the channel on PR merge 1`] = `
{
  "closedChannels": [
    {
      "channel": "pr-test-proj-test-repository-1",
    },
  ],
  "createdChannels": [],
  "invitesToChannels": [],
  "lookedUpUsers": [],
  "sentMessages": [
    {
      "channel": "pr-test-proj-test-repository-1",
      "text": "The pull request was merged by test.author. Well done, thank you all.",
    },
  ],
  "setChannelTopics": [],
}
`;
